### PR TITLE
[15565] Feature: new on_unacknowledged_sample_removed callback in DataWriterListener

### PR DESCRIPTION
**Pull request by JLBuenoLopez-eProsima**
## Description

This PR includes the new `DataWriterListener::on_unacknowledged_sample_removed` callback tests and implementation.
Documentation PR: eProsima/Fast-DDS-docs#439.

## Contributor Checklist
- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- [x] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
- [x] Any new/modified methods have been properly documented using Doxygen. 
- Changes are **NOT** ABI compatible. 
- [x] Changes are API compatible. 
- [x] Documentation builds and tests pass locally. 
- [x] New feature has been added to the `versions.md` file (if applicable).
- [x] New feature has been documented/Current behavior is correctly described in the documentation. 

**N/A** Applicable backports have been included in the description.

## Reviewer Checklist
- [x] Check contributor checklist is correct.
- [x] Check CI results: changes do not issue any warning.
- [x] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

**Comment by JLBuenoLopez-eProsima** [comment]
@richiprosima please test this

File: src/cpp/rtps/writer/StatelessWriter.cpp [hunk 1]
```diff
@@ -385,11 +385,27 @@ bool StatelessWriter::change_removed_by_history(
     return true;
 }
 
+bool StatelessWriter::has_been_fully_delivered(
+        const SequenceNumber_t& seq_num) const
+{
+    if (getMatchedReadersSize() > 0)
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply src/cpp/rtps/writer/StatelessWriter.cpp:400]: A test case has been included in commit 3eba6b8 and a fix in b89b1b8

File: src/cpp/fastdds/publisher/DataWriterImpl.cpp [hunk 2]
```diff
@@ -151,8 +151,16 @@ DataWriterImpl::DataWriterImpl(
     , type_(type)
     , topic_(topic)
     , qos_(&qos == &DATAWRITER_QOS_DEFAULT ? publisher_->get_default_datawriter_qos() : qos)
-    , history_(get_topic_attributes(qos_, *topic_, type_), type_->m_typeSize, qos_.endpoint().history_memory_policy)
     , listener_(listen)
+    , history_(get_topic_attributes(qos_, *topic_, type_), type_->m_typeSize, qos_.endpoint().history_memory_policy,
+            [&](
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply src/cpp/fastdds/publisher/DataWriterImpl.cpp]: Applied in 1c020b6

File: src/cpp/rtps/writer/StatefulWriter.cpp [hunk 3]
```diff
@@ -1322,6 +1322,28 @@ bool StatefulWriter::matched_reader_lookup(
                    );
 }
 
+bool StatefulWriter::has_been_fully_delivered(
+        const SequenceNumber_t& seq_num) const
+{
+    std::lock_guard<RecursiveTimedMutex> guard(mp_mutex);
+    bool found = false;
+    // Sequence number has not been generated by this WriterHistory.
+    if (seq_num > (mp_history->next_sequence_number() - 1))
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply src/cpp/rtps/writer/StatefulWriter.cpp]: Applied in 1c020b6

**Review by MiguelCompany — APPROVED** [review]

File: include/fastdds/rtps/writer/RTPSWriter.h [hunk 4]
```diff
@@ -194,6 +194,20 @@ class RTPSWriter
      */
     RTPS_DllAPI virtual const fastdds::rtps::IReaderDataFilter* reader_data_filter() const = 0;
 
+    /**
+     * @brief Check if a specific change has been delivered to the transport layer of every matched remote RTPSReader
+     * at least once.
+     *
+     * @param change Pointer to the CacheChange to check.
```
Comment (MiguelCompany) [inline include/fastdds/rtps/writer/RTPSWriter.h:201]: This is making the documentation build fail.

```suggestion
     * @param seq_num Sequence number of the change to check.
```

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

**Review by MiguelCompany — APPROVED** [review]

**Pull request merged**